## Setup

### Importació de les dades.

In [3]:
library(ElemStatLearn)
data(spam)
wine<-read.csv('wine.csv',header=FALSE)
colnames(wine)<-c("Type","Alcohol","Malic","Ash","Alcalinity","Magnesium","Phenols","Flavonoids","Nonflavonoids","Proanthocyanins","Color","Hue","Dilution","Proline")
wine$Type <- as.factor(wine$Type)

Warning message:
"package 'ElemStatLearn' was built under R version 3.3.3"

### Llibreries

In [65]:
library(class)

## Classificació de Spam

Descripció de les dades **TODO**

In [66]:
nrow(spam)

[1] 4601

Veiem que el dataset és bastant gran aixi que utilitzarem com a model de predicció k-fold cross-validation. 
No utilitcem Leave-one-out per que seria poc eficient fer les 4601 fileres, tampoc utilitzarem boostrap ja que no tenim necessitat de ampliar la mostra.

Utilitzarem un k de 10

In [67]:
spam<-spam[sample(nrow(spam)),]

### Classificació per  K-nn
Utilitzarem la funció *knn* que trobem al paquet *class*.
Amb la quantitat de dades que hi han testejarem amb una k= 10 i 20.

In [68]:
k<-10

kfold<-10
totalprecisio.email<-0
totalprecisio.spam<-0
for (i in 1:kfold){
    ITest<-((nrow(spam)/kfold)*(i-1)+1):((nrow(spam)/kfold)*i)
    spam.train<-spam[-ITest,-ncol(spam)]
    spam.test<-spam[ITest,-ncol(spam)]
    spam.Ytrain<-spam[-ITest,ncol(spam)]
    spam.Ytest<-spam[ITest,ncol(spam)]
    spam.Yhat<-knn(spam.train, spam.test, spam.Ytrain, k = k)
    C<-table("True"=spam.Ytest,"Predicted"=spam.Yhat)
    cat('\nMatriu confusio\n')
    print(C)
    precisio.email<-C[1]/(C[1]+C[3])
    precisio.spam<-C[4]/(C[2]+C[4])
    cat('preciso email ',+precisio.email)
    cat('\nprecisio spam ',precisio.spam)
    cat('\n')
    totalprecisio.email<-totalprecisio.email + precisio.email
    totalprecisio.spam<-totalprecisio.spam + precisio.spam
}
cat('\npreciso email mitjana ',totalprecisio.email/kfold)
cat('\nprecisio spam mitjana ',totalprecisio.spam/kfold)


Matriu confusio
       Predicted
True    email spam
  email   224   36
  spam     50  150
preciso email  0.8615385
precisio spam  0.75

Matriu confusio
       Predicted
True    email spam
  email   255   38
  spam     53  114
preciso email  0.8703072
precisio spam  0.6826347

Matriu confusio
       Predicted
True    email spam
  email   228   62
  spam     47  123
preciso email  0.7862069
precisio spam  0.7235294

Matriu confusio
       Predicted
True    email spam
  email   245   36
  spam     52  127
preciso email  0.8718861
precisio spam  0.7094972

Matriu confusio
       Predicted
True    email spam
  email   224   46
  spam     38  152
preciso email  0.8296296
precisio spam  0.8

Matriu confusio
       Predicted
True    email spam
  email   239   55
  spam     56  110
preciso email  0.8129252
precisio spam  0.6626506

Matriu confusio
       Predicted
True    email spam
  email   230   44
  spam     48  138
preciso email  0.8394161
precisio spam  0.7419355

Matriu confusio
       

In [69]:
k<-20

kfold<-10
totalprecisio.email<-0
totalprecisio.spam<-0
for (i in 1:kfold){
    ITest<-((nrow(spam)/kfold)*(i-1)+1):((nrow(spam)/kfold)*i)
    spam.train<-spam[-ITest,-ncol(spam)]
    spam.test<-spam[ITest,-ncol(spam)]
    spam.Ytrain<-spam[-ITest,ncol(spam)]
    spam.Ytest<-spam[ITest,ncol(spam)]
    spam.Yhat<-knn(spam.train, spam.test, spam.Ytrain, k = k)
    C<-table("True"=spam.Ytest,"Predicted"=spam.Yhat)
    cat('\nMatriu confusio\n')
    print(C)
    precisio.email<-C[1]/(C[1]+C[3])
    precisio.spam<-C[4]/(C[2]+C[4])
    cat('preciso email ',+precisio.email)
    cat('\nprecisio spam ',precisio.spam)
    cat('\n')
    totalprecisio.email<-totalprecisio.email + precisio.email
    totalprecisio.spam<-totalprecisio.spam + precisio.spam
}
cat('\npreciso email mitjana ',totalprecisio.email/kfold)
cat('\nprecisio spam mitjana ',totalprecisio.spam/kfold)


Matriu confusio
       Predicted
True    email spam
  email   221   39
  spam     61  139
preciso email  0.85
precisio spam  0.695

Matriu confusio
       Predicted
True    email spam
  email   244   49
  spam     53  114
preciso email  0.8327645
precisio spam  0.6826347

Matriu confusio
       Predicted
True    email spam
  email   221   69
  spam     51  119
preciso email  0.762069
precisio spam  0.7

Matriu confusio
       Predicted
True    email spam
  email   242   39
  spam     64  115
preciso email  0.86121
precisio spam  0.6424581

Matriu confusio
       Predicted
True    email spam
  email   214   56
  spam     42  148
preciso email  0.7925926
precisio spam  0.7789474

Matriu confusio
       Predicted
True    email spam
  email   242   52
  spam     63  103
preciso email  0.8231293
precisio spam  0.6204819

Matriu confusio
       Predicted
True    email spam
  email   227   47
  spam     49  137
preciso email  0.8284672
precisio spam  0.7365591

Matriu confusio
       Predict

Veiem que optenim mes precisió en k = 10 que en k = 20. Encara que la precisió de predicció de spam és més baixa que amb email.

### Regresió logística

Hold-out

In [24]:
x <- c(spam = 0, email = 1)
spam.categorize<-spam[-nrow(spam)]
spam.categorize$spam <- x[spam$spam]

In [25]:
ITest<-sample(1:nrow(spam),nrow(spam)/10,replace=FALSE)
spam.train<-spam[-ITest,]
spam.test<-spam[ITest,]
spam.logist<-glm(spam~.,data=spam.train,family=binomial)

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"

In [32]:
summary(spam.logist)
spam.pred<-predict(spam.logist,newdata=spam.test, type="response")
spam.pred.crisp<-1*(spam.pred>=0.5)
C<-table("True"=spam.test$spam,"Predicted"=spam.pred.crisp)
C


Call:
glm(formula = spam ~ ., family = binomial, data = spam.train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-4.1312  -0.2143   0.0000   0.1044   5.1887  

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.596e+00  1.486e-01 -10.737  < 2e-16 ***
A.1         -5.423e-01  2.530e-01  -2.143 0.032111 *  
A.2         -1.497e-01  7.534e-02  -1.987 0.046952 *  
A.3          1.070e-01  1.145e-01   0.934 0.350057    
A.4          2.833e+00  1.718e+00   1.649 0.099083 .  
A.5          5.846e-01  1.075e-01   5.439 5.36e-08 ***
A.6          8.709e-01  2.649e-01   3.288 0.001009 ** 
A.7          2.097e+00  3.354e-01   6.253 4.03e-10 ***
A.8          4.752e-01  1.556e-01   3.053 0.002265 ** 
A.9          5.258e-01  2.847e-01   1.847 0.064750 .  
A.10         1.212e-01  7.271e-02   1.668 0.095413 .  
A.11        -1.250e-01  3.110e-01  -0.402 0.687814    
A.12        -1.589e-01  7.861e-02  -2.021 0.043313 *  
A.13        -2.991e-02  2.541e-0

       Predicted
True      0   1
  email 260  19
  spam   20 161

In [33]:
step(spam.logist)

Start:  AIC=1742.84
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.11 + A.12 + A.13 + A.14 + A.15 + A.16 + A.17 + A.18 + 
    A.19 + A.20 + A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + 
    A.28 + A.29 + A.30 + A.31 + A.32 + A.33 + A.34 + A.35 + A.36 + 
    A.37 + A.38 + A.39 + A.40 + A.41 + A.42 + A.43 + A.44 + A.45 + 
    A.46 + A.47 + A.48 + A.49 + A.50 + A.51 + A.52 + A.53 + A.54 + 
    A.55 + A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
- A.13  1   1626.9 1740.9
- A.37  1   1626.9 1740.9
- A.55  1   1626.9 1740.9
- A.11  1   1627.0 1741.0
- A.31  1   1627.1 1741.1
- A.30  1   1627.4 1741.4
- A.34  1   1627.6 1741.6
- A.3   1   1627.7 1741.7
- A.50  1   1627.7 1741.7
- A.14  1   1627.8 1741.8
- A.51  1   1627.8 1741.8
- A.40  1   1627.9 1741.9
- A.18  1   1627.9 1741.9
- A.32  1   1628.2 1742.2
- A.22  1   1628.5 1742.5
<none>      1626.8 1742.8
- A.15  1   1629.5 1743.5
- A.10  1   1629.8 1743.8
- A.9   1   1630.4 1744.4
- A.43  1   1630.9 1744.9
- A.38  1   1631.0 1745.0
- A.12  1   1631.3 1745.3
- A.47  1   1631.4 1745.4
- A.19  1   1631.4 1745.4
- A.1   1   1632.0 1746.0
- A.35  1   1632.8 1746.8
- A.39  1   1632.8 1746.8
- A.36  1   1633.8 1747.8
- A.2   1   1634.4 1748.4
- A.28  1   1634.4 1748.4
- A.54  1   1635.0 1749.0
- A.26  1   1635.1 1749.1
- A.24  1   1636.8 1750.8
- A.4   1   1637.0 1751.0
- A.33  1   1637.0 1751.0
- A.20  1   1637.2 1751.2
- A.8   1   1639.2 1753.2
- A.56  1   

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1740.86
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.11 + A.12 + A.14 + A.15 + A.16 + A.17 + A.18 + A.19 + 
    A.20 + A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + 
    A.29 + A.30 + A.31 + A.32 + A.33 + A.34 + A.35 + A.36 + A.37 + 
    A.38 + A.39 + A.40 + A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + 
    A.47 + A.48 + A.49 + A.50 + A.51 + A.52 + A.53 + A.54 + A.55 + 
    A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
- A.37  1   1626.9 1738.9
- A.55  1   1627.0 1739.0
- A.11  1   1627.0 1739.0
- A.31  1   1627.1 1739.1
- A.30  1   1627.4 1739.4
- A.34  1   1627.6 1739.6
- A.3   1   1627.7 1739.7
- A.50  1   1627.7 1739.7
- A.51  1   1627.8 1739.8
- A.14  1   1627.8 1739.8
- A.40  1   1627.9 1739.9
- A.18  1   1627.9 1739.9
- A.32  1   1628.3 1740.3
- A.22  1   1628.5 1740.5
<none>      1626.9 1740.9
- A.15  1   1629.5 1741.5
- A.10  1   1629.8 1741.8
- A.9   1   1630.5 1742.5
- A.43  1   1630.9 1742.9
- A.38  1   1631.0 1743.0
- A.12  1   1631.3 1743.3
- A.47  1   1631.4 1743.4
- A.19  1   1631.4 1743.4
- A.1   1   1632.0 1744.0
- A.35  1   1632.8 1744.8
- A.39  1   1632.9 1744.9
- A.36  1   1633.9 1745.9
- A.2   1   1634.4 1746.4
- A.28  1   1634.4 1746.4
- A.54  1   1635.0 1747.0
- A.26  1   1635.1 1747.1
- A.24  1   1636.9 1748.9
- A.33  1   1637.0 1749.0
- A.4   1   1637.0 1749.0
- A.20  1   1637.2 1749.2
- A.8   1   1639.2 1751.2
- A.56  1   1639.9 1751.9
- A.29  1   

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1738.88
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.11 + A.12 + A.14 + A.15 + A.16 + A.17 + A.18 + A.19 + 
    A.20 + A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + 
    A.29 + A.30 + A.31 + A.32 + A.33 + A.34 + A.35 + A.36 + A.38 + 
    A.39 + A.40 + A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + A.47 + 
    A.48 + A.49 + A.50 + A.51 + A.52 + A.53 + A.54 + A.55 + A.56 + 
    A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
- A.55  1   1627.0 1737.0
- A.11  1   1627.0 1737.0
- A.31  1   1627.1 1737.1
- A.30  1   1627.4 1737.4
- A.34  1   1627.6 1737.6
- A.3   1   1627.7 1737.7
- A.50  1   1627.8 1737.8
- A.51  1   1627.8 1737.8
- A.14  1   1627.8 1737.8
- A.40  1   1627.9 1737.9
- A.18  1   1627.9 1737.9
- A.32  1   1628.3 1738.3
- A.22  1   1628.6 1738.6
<none>      1626.9 1738.9
- A.15  1   1629.5 1739.5
- A.10  1   1629.8 1739.8
- A.9   1   1630.5 1740.5
- A.43  1   1631.0 1741.0
- A.38  1   1631.0 1741.0
- A.12  1   1631.3 1741.3
- A.47  1   1631.4 1741.4
- A.19  1   1631.5 1741.5
- A.1   1   1632.1 1742.1
- A.35  1   1632.9 1742.9
- A.39  1   1632.9 1742.9
- A.36  1   1633.9 1743.9
- A.2   1   1634.4 1744.4
- A.28  1   1634.5 1744.5
- A.54  1   1635.0 1745.0
- A.26  1   1635.2 1745.2
- A.24  1   1636.9 1746.9
- A.4   1   1637.0 1747.0
- A.33  1   1637.1 1747.1
- A.20  1   1637.3 1747.3
- A.8   1   1639.3 1749.3
- A.56  1   1640.0 1750.0
- A.29  1   1640.3 1750.3
- A.6   1   

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1736.97
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.11 + A.12 + A.14 + A.15 + A.16 + A.17 + A.18 + A.19 + 
    A.20 + A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + 
    A.29 + A.30 + A.31 + A.32 + A.33 + A.34 + A.35 + A.36 + A.38 + 
    A.39 + A.40 + A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + A.47 + 
    A.48 + A.49 + A.50 + A.51 + A.52 + A.53 + A.54 + A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance   AIC
- A.11  1     1627  1735
- A.31  1     1627  1735
- A.30  1     1627  1735
- A.34  1     1628  1736
- A.3   1     1628  1736
- A.50  1     1628  1736
- A.14  1     1628  1736
- A.51  1     1628  1736
- A.18  1     1628  1736
- A.32  1     1628  1736
- A.22  1     1629  1737
<none>        1627  1737
- A.15  1     1630  1738
- A.10  1     1630  1738
- A.9   1     1631  1739
- A.43  1     1631  1739
- A.38  1     1631  1739
- A.47  1     1631  1739
- A.12  1     1631  1739
- A.19  1     1632  1740
- A.1   1     1632  1740
- A.39  1     1633  1741
- A.35  1     1633  1741
- A.36  1     1634  1742
- A.2   1     1635  1743
- A.28  1     1635  1743
- A.54  1     1635  1743
- A.26  1     1635  1743
- A.24  1     1637  1745
- A.4   1     1637  1745
- A.33  1     1637  1745
- A.20  1     1637  1745
- A.8   1     1639  1747
- A.29  1     1640  1748
- A.6   1     1641  1749
- A.44  1     1642  1750
- A.48  1     1643  1751
- A.49  1     1643  1751
- A.17  1     1645  1753


Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1735.15
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.12 + A.14 + A.15 + A.16 + A.17 + A.18 + A.19 + A.20 + 
    A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + 
    A.30 + A.31 + A.32 + A.33 + A.34 + A.35 + A.36 + A.38 + A.39 + 
    A.40 + A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + A.47 + A.48 + 
    A.49 + A.50 + A.51 + A.52 + A.53 + A.54 + A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance   AIC
- A.31  1     1627  1733
- A.30  1     1628  1734
- A.34  1     1628  1734
- A.50  1     1628  1734
- A.14  1     1628  1734
- A.3   1     1628  1734
- A.51  1     1628  1734
- A.18  1     1628  1734
- A.32  1     1629  1735
- A.22  1     1629  1735
<none>        1627  1735
- A.15  1     1630  1736
- A.10  1     1630  1736
- A.9   1     1631  1737
- A.43  1     1631  1737
- A.38  1     1631  1737
- A.47  1     1632  1738
- A.19  1     1632  1738
- A.12  1     1632  1738
- A.39  1     1633  1739
- A.35  1     1633  1739
- A.1   1     1634  1740
- A.36  1     1634  1740
- A.2   1     1635  1741
- A.28  1     1635  1741
- A.26  1     1635  1741
- A.54  1     1635  1741
- A.4   1     1637  1743
- A.20  1     1637  1743
- A.24  1     1637  1743
- A.33  1     1637  1743
- A.8   1     1639  1745
- A.29  1     1640  1746
- A.6   1     1641  1747
- A.44  1     1642  1748
- A.48  1     1643  1749
- A.49  1     1643  1749
- A.17  1     1645  1751
- A.41  1     1649  1755


Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1733.39
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.12 + A.14 + A.15 + A.16 + A.17 + A.18 + A.19 + A.20 + 
    A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + 
    A.30 + A.32 + A.33 + A.34 + A.35 + A.36 + A.38 + A.39 + A.40 + 
    A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + A.47 + A.48 + A.49 + 
    A.50 + A.51 + A.52 + A.53 + A.54 + A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
- A.30  1   1627.9 1731.9
- A.34  1   1628.2 1732.2
- A.50  1   1628.2 1732.2
- A.14  1   1628.3 1732.3
- A.3   1   1628.3 1732.3
- A.51  1   1628.3 1732.3
- A.18  1   1628.4 1732.4
- A.40  1   1628.4 1732.4
- A.32  1   1628.8 1732.8
- A.22  1   1629.1 1733.1
<none>      1627.4 1733.4
- A.15  1   1630.0 1734.0
- A.10  1   1630.3 1734.3
- A.9   1   1631.0 1735.0
- A.43  1   1631.5 1735.5
- A.38  1   1631.6 1735.6
- A.47  1   1631.8 1735.8
- A.19  1   1632.0 1736.0
- A.12  1   1632.1 1736.1
- A.39  1   1633.4 1737.4
- A.35  1   1633.7 1737.7
- A.1   1   1634.2 1738.2
- A.36  1   1634.4 1738.4
- A.2   1   1634.9 1738.9
- A.28  1   1635.0 1739.0
- A.26  1   1635.5 1739.5
- A.54  1   1635.6 1739.6
- A.4   1   1637.6 1741.6
- A.20  1   1637.6 1741.6
- A.24  1   1637.6 1741.6
- A.33  1   1637.7 1741.7
- A.8   1   1639.7 1743.7
- A.29  1   1640.6 1744.6
- A.6   1   1641.1 1745.1
- A.48  1   1643.1 1747.1
- A.49  1   1643.5 1747.5
- A.44  1   1645.4 1749.4
- A.17  1   

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1731.93
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.12 + A.14 + A.15 + A.16 + A.17 + A.18 + A.19 + A.20 + 
    A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + 
    A.32 + A.33 + A.34 + A.35 + A.36 + A.38 + A.39 + A.40 + A.41 + 
    A.42 + A.43 + A.44 + A.45 + A.46 + A.47 + A.48 + A.49 + A.50 + 
    A.51 + A.52 + A.53 + A.54 + A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
- A.34  1   1628.7 1730.7
- A.50  1   1628.8 1730.8
- A.3   1   1628.8 1730.8
- A.14  1   1628.8 1730.8
- A.51  1   1628.8 1730.8
- A.40  1   1628.9 1730.9
- A.18  1   1629.0 1731.0
- A.32  1   1629.3 1731.3
- A.22  1   1629.7 1731.7
<none>      1627.9 1731.9
- A.15  1   1630.5 1732.5
- A.10  1   1630.8 1732.8
- A.9   1   1631.4 1733.4
- A.43  1   1632.0 1734.0
- A.38  1   1632.1 1734.1
- A.47  1   1632.3 1734.3
- A.12  1   1632.5 1734.5
- A.19  1   1632.5 1734.5
- A.39  1   1634.0 1736.0
- A.35  1   1634.2 1736.2
- A.1   1   1634.7 1736.7
- A.36  1   1635.0 1737.0
- A.2   1   1635.4 1737.4
- A.28  1   1635.5 1737.5
- A.26  1   1636.1 1738.1
- A.54  1   1636.2 1738.2
- A.33  1   1638.1 1740.1
- A.4   1   1638.1 1740.1
- A.24  1   1638.2 1740.2
- A.20  1   1638.3 1740.3
- A.8   1   1640.3 1742.3
- A.29  1   1641.3 1743.3
- A.6   1   1641.4 1743.4
- A.48  1   1643.7 1745.7
- A.49  1   1644.0 1746.0
- A.44  1   1646.1 1748.1
- A.17  1   1646.3 1748.3
- A.41  1   

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1730.69
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.12 + A.14 + A.15 + A.16 + A.17 + A.18 + A.19 + A.20 + 
    A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + 
    A.32 + A.33 + A.35 + A.36 + A.38 + A.39 + A.40 + A.41 + A.42 + 
    A.43 + A.44 + A.45 + A.46 + A.47 + A.48 + A.49 + A.50 + A.51 + 
    A.52 + A.53 + A.54 + A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
- A.50  1   1629.5 1729.5
- A.3   1   1629.6 1729.6
- A.14  1   1629.6 1729.6
- A.51  1   1629.6 1729.6
- A.40  1   1629.7 1729.7
- A.18  1   1629.7 1729.7
- A.32  1   1630.3 1730.3
- A.22  1   1630.5 1730.5
<none>      1628.7 1730.7
- A.15  1   1631.2 1731.2
- A.10  1   1631.7 1731.7
- A.9   1   1632.1 1732.1
- A.43  1   1632.8 1732.8
- A.38  1   1632.8 1732.8
- A.47  1   1633.1 1733.1
- A.19  1   1633.2 1733.2
- A.12  1   1633.4 1733.4
- A.39  1   1634.7 1734.7
- A.35  1   1635.0 1735.0
- A.1   1   1635.3 1735.3
- A.36  1   1635.8 1735.8
- A.2   1   1636.2 1736.2
- A.28  1   1636.3 1736.3
- A.26  1   1636.8 1736.8
- A.54  1   1636.9 1736.9
- A.33  1   1638.9 1738.9
- A.4   1   1638.9 1738.9
- A.24  1   1639.0 1739.0
- A.20  1   1639.0 1739.0
- A.8   1   1641.1 1741.1
- A.29  1   1642.1 1742.1
- A.6   1   1642.1 1742.1
- A.48  1   1644.4 1744.4
- A.49  1   1644.8 1744.8
- A.44  1   1646.9 1746.9
- A.17  1   1647.0 1747.0
- A.41  1   1650.5 1750.5
- A.21  1   

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1729.55
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.12 + A.14 + A.15 + A.16 + A.17 + A.18 + A.19 + A.20 + 
    A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + 
    A.32 + A.33 + A.35 + A.36 + A.38 + A.39 + A.40 + A.41 + A.42 + 
    A.43 + A.44 + A.45 + A.46 + A.47 + A.48 + A.49 + A.51 + A.52 + 
    A.53 + A.54 + A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
- A.3   1   1630.4 1728.4
- A.51  1   1630.5 1728.5
- A.14  1   1630.5 1728.5
- A.40  1   1630.5 1728.5
- A.18  1   1630.6 1728.6
- A.32  1   1631.1 1729.1
- A.22  1   1631.5 1729.5
<none>      1629.5 1729.5
- A.15  1   1632.1 1730.1
- A.10  1   1632.6 1730.6
- A.9   1   1632.9 1730.9
- A.43  1   1633.6 1731.6
- A.38  1   1633.6 1731.6
- A.47  1   1634.0 1732.0
- A.12  1   1634.1 1732.1
- A.19  1   1634.5 1732.5
- A.39  1   1635.5 1733.5
- A.35  1   1635.8 1733.8
- A.1   1   1636.1 1734.1
- A.28  1   1636.3 1734.3
- A.36  1   1636.7 1734.7
- A.2   1   1636.8 1734.8
- A.26  1   1637.7 1735.7
- A.54  1   1637.7 1735.7
- A.33  1   1639.7 1737.7
- A.24  1   1639.7 1737.7
- A.20  1   1639.8 1737.8
- A.4   1   1639.8 1737.8
- A.8   1   1642.2 1740.2
- A.6   1   1642.8 1740.8
- A.29  1   1643.0 1741.0
- A.48  1   1645.3 1743.3
- A.49  1   1645.7 1743.7
- A.44  1   1647.6 1745.6
- A.17  1   1647.8 1745.8
- A.41  1   1651.3 1749.3
- A.57  1   1655.3 1753.3
- A.21  1   

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1728.42
spam ~ A.1 + A.2 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + A.10 + 
    A.12 + A.14 + A.15 + A.16 + A.17 + A.18 + A.19 + A.20 + A.21 + 
    A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + A.32 + 
    A.33 + A.35 + A.36 + A.38 + A.39 + A.40 + A.41 + A.42 + A.43 + 
    A.44 + A.45 + A.46 + A.47 + A.48 + A.49 + A.51 + A.52 + A.53 + 
    A.54 + A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
- A.14  1   1631.3 1727.3
- A.51  1   1631.4 1727.4
- A.40  1   1631.4 1727.4
- A.18  1   1631.6 1727.6
- A.32  1   1632.1 1728.1
- A.22  1   1632.3 1728.3
<none>      1630.4 1728.4
- A.15  1   1633.1 1729.1
- A.10  1   1633.3 1729.3
- A.9   1   1633.9 1729.9
- A.43  1   1634.5 1730.5
- A.38  1   1634.6 1730.6
- A.47  1   1634.6 1730.6
- A.12  1   1634.9 1730.9
- A.19  1   1635.8 1731.8
- A.39  1   1636.3 1732.3
- A.35  1   1636.8 1732.8
- A.28  1   1637.1 1733.1
- A.1   1   1637.1 1733.1
- A.36  1   1637.4 1733.4
- A.2   1   1637.8 1733.8
- A.54  1   1638.6 1734.6
- A.26  1   1638.6 1734.6
- A.24  1   1640.5 1736.5
- A.4   1   1640.6 1736.6
- A.20  1   1640.7 1736.7
- A.33  1   1640.8 1736.8
- A.8   1   1642.8 1738.8
- A.29  1   1643.5 1739.5
- A.6   1   1643.5 1739.5
- A.48  1   1646.3 1742.3
- A.49  1   1646.7 1742.7
- A.17  1   1648.7 1744.7
- A.44  1   1648.8 1744.8
- A.41  1   1652.3 1748.3
- A.57  1   1656.2 1752.2
- A.21  1   1656.6 1752.6
- A.52  1   

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1727.26
spam ~ A.1 + A.2 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + A.10 + 
    A.12 + A.15 + A.16 + A.17 + A.18 + A.19 + A.20 + A.21 + A.22 + 
    A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + A.32 + A.33 + 
    A.35 + A.36 + A.38 + A.39 + A.40 + A.41 + A.42 + A.43 + A.44 + 
    A.45 + A.46 + A.47 + A.48 + A.49 + A.51 + A.52 + A.53 + A.54 + 
    A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
- A.51  1   1632.2 1726.2
- A.40  1   1632.3 1726.3
- A.18  1   1632.4 1726.4
- A.32  1   1632.9 1726.9
- A.22  1   1633.1 1727.1
<none>      1631.3 1727.3
- A.15  1   1634.0 1728.0
- A.10  1   1634.2 1728.2
- A.9   1   1634.7 1728.7
- A.43  1   1635.3 1729.3
- A.38  1   1635.4 1729.4
- A.47  1   1635.5 1729.5
- A.12  1   1635.8 1729.8
- A.19  1   1636.6 1730.6
- A.39  1   1637.3 1731.3
- A.35  1   1637.6 1731.6
- A.28  1   1637.9 1731.9
- A.1   1   1638.0 1732.0
- A.36  1   1638.8 1732.8
- A.2   1   1638.8 1732.8
- A.54  1   1639.4 1733.4
- A.26  1   1639.7 1733.7
- A.24  1   1641.2 1735.2
- A.4   1   1641.4 1735.4
- A.20  1   1641.6 1735.6
- A.33  1   1641.7 1735.7
- A.8   1   1643.6 1737.6
- A.29  1   1644.2 1738.2
- A.6   1   1644.2 1738.2
- A.48  1   1647.3 1741.3
- A.49  1   1647.6 1741.6
- A.17  1   1649.3 1743.3
- A.44  1   1649.6 1743.6
- A.41  1   1653.2 1747.2
- A.57  1   1657.1 1751.1
- A.21  1   1657.2 1751.2
- A.52  1   1660.7 1754.7
- A.56  1   

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1726.25
spam ~ A.1 + A.2 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + A.10 + 
    A.12 + A.15 + A.16 + A.17 + A.18 + A.19 + A.20 + A.21 + A.22 + 
    A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + A.32 + A.33 + 
    A.35 + A.36 + A.38 + A.39 + A.40 + A.41 + A.42 + A.43 + A.44 + 
    A.45 + A.46 + A.47 + A.48 + A.49 + A.52 + A.53 + A.54 + A.56 + 
    A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
- A.40  1   1633.2 1725.2
- A.18  1   1633.4 1725.4
- A.32  1   1633.9 1725.9
- A.22  1   1634.1 1726.1
<none>      1632.2 1726.2
- A.15  1   1635.0 1727.0
- A.10  1   1635.2 1727.2
- A.9   1   1635.7 1727.7
- A.43  1   1636.3 1728.3
- A.38  1   1636.4 1728.4
- A.47  1   1636.4 1728.4
- A.12  1   1636.7 1728.7
- A.19  1   1637.8 1729.8
- A.39  1   1638.2 1730.2
- A.35  1   1638.5 1730.5
- A.28  1   1638.9 1730.9
- A.1   1   1639.0 1731.0
- A.2   1   1639.7 1731.7
- A.36  1   1639.9 1731.9
- A.54  1   1640.4 1732.4
- A.26  1   1640.7 1732.7
- A.24  1   1642.2 1734.2
- A.4   1   1642.5 1734.5
- A.20  1   1642.6 1734.6
- A.33  1   1642.8 1734.8
- A.8   1   1644.7 1736.7
- A.6   1   1645.1 1737.1
- A.29  1   1645.2 1737.2
- A.48  1   1648.2 1740.2
- A.49  1   1648.5 1740.5
- A.17  1   1650.5 1742.5
- A.44  1   1650.5 1742.5
- A.41  1   1654.2 1746.2
- A.57  1   1657.8 1749.8
- A.21  1   1658.2 1750.2
- A.52  1   1661.9 1753.9
- A.56  1   1665.1 1757.1
- A.23  1   

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1725.25
spam ~ A.1 + A.2 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + A.10 + 
    A.12 + A.15 + A.16 + A.17 + A.18 + A.19 + A.20 + A.21 + A.22 + 
    A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + A.32 + A.33 + 
    A.35 + A.36 + A.38 + A.39 + A.41 + A.42 + A.43 + A.44 + A.45 + 
    A.46 + A.47 + A.48 + A.49 + A.52 + A.53 + A.54 + A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
- A.18  1   1634.4 1724.4
- A.32  1   1634.9 1724.9
<none>      1633.2 1725.2
- A.22  1   1635.3 1725.3
- A.15  1   1635.9 1725.9
- A.10  1   1636.0 1726.0
- A.9   1   1636.5 1726.5
- A.43  1   1637.3 1727.3
- A.38  1   1637.3 1727.3
- A.47  1   1637.5 1727.5
- A.12  1   1637.5 1727.5
- A.19  1   1639.0 1729.0
- A.39  1   1639.2 1729.2
- A.35  1   1639.5 1729.5
- A.1   1   1639.9 1729.9
- A.28  1   1639.9 1729.9
- A.2   1   1640.6 1730.6
- A.54  1   1640.7 1730.7
- A.36  1   1640.8 1730.8
- A.26  1   1641.7 1731.7
- A.24  1   1643.3 1733.3
- A.4   1   1643.5 1733.5
- A.20  1   1643.6 1733.6
- A.33  1   1643.7 1733.7
- A.8   1   1645.7 1735.7
- A.29  1   1646.1 1736.1
- A.6   1   1646.1 1736.1
- A.48  1   1649.3 1739.3
- A.49  1   1649.5 1739.5
- A.44  1   1651.1 1741.1
- A.17  1   1651.2 1741.2
- A.41  1   1655.1 1745.1
- A.57  1   1658.9 1748.9
- A.21  1   1658.9 1748.9
- A.52  1   1663.0 1753.0
- A.56  1   1666.2 1756.2
- A.23  1   1667.4 1757.4
- A.5   1   

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1724.36
spam ~ A.1 + A.2 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + A.10 + 
    A.12 + A.15 + A.16 + A.17 + A.19 + A.20 + A.21 + A.22 + A.23 + 
    A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + A.32 + A.33 + A.35 + 
    A.36 + A.38 + A.39 + A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + 
    A.47 + A.48 + A.49 + A.52 + A.53 + A.54 + A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
- A.32  1   1636.0 1724.0
<none>      1634.4 1724.4
- A.22  1   1636.4 1724.4
- A.10  1   1637.1 1725.1
- A.15  1   1637.5 1725.5
- A.9   1   1637.6 1725.6
- A.47  1   1638.3 1726.3
- A.43  1   1638.4 1726.4
- A.38  1   1638.6 1726.6
- A.12  1   1638.8 1726.8
- A.39  1   1640.5 1728.5
- A.19  1   1640.6 1728.6
- A.35  1   1640.8 1728.8
- A.28  1   1641.1 1729.1
- A.1   1   1641.4 1729.4
- A.2   1   1641.5 1729.5
- A.54  1   1641.6 1729.6
- A.36  1   1641.8 1729.8
- A.26  1   1642.8 1730.8
- A.24  1   1644.2 1732.2
- A.4   1   1644.5 1732.5
- A.20  1   1644.6 1732.6
- A.33  1   1645.0 1733.0
- A.8   1   1646.8 1734.8
- A.6   1   1647.0 1735.0
- A.29  1   1647.4 1735.4
- A.48  1   1650.5 1738.5
- A.49  1   1650.8 1738.8
- A.44  1   1652.4 1740.4
- A.17  1   1652.8 1740.8
- A.41  1   1656.1 1744.1
- A.57  1   1659.7 1747.7
- A.21  1   1660.5 1748.5
- A.52  1   1663.9 1751.9
- A.23  1   1668.1 1756.1
- A.56  1   1668.2 1756.2
- A.5   1   1673.6 1761.6
- A.42  1   

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1724.02
spam ~ A.1 + A.2 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + A.10 + 
    A.12 + A.15 + A.16 + A.17 + A.19 + A.20 + A.21 + A.22 + A.23 + 
    A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + A.33 + A.35 + A.36 + 
    A.38 + A.39 + A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + A.47 + 
    A.48 + A.49 + A.52 + A.53 + A.54 + A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
- A.22  1   1638.0 1724.0
<none>      1636.0 1724.0
- A.10  1   1638.7 1724.7
- A.15  1   1639.1 1725.1
- A.9   1   1639.1 1725.1
- A.47  1   1640.0 1726.0
- A.43  1   1640.1 1726.1
- A.38  1   1640.3 1726.3
- A.12  1   1640.6 1726.6
- A.19  1   1642.0 1728.0
- A.39  1   1642.2 1728.2
- A.35  1   1642.4 1728.4
- A.28  1   1642.8 1728.8
- A.2   1   1643.1 1729.1
- A.1   1   1643.2 1729.2
- A.54  1   1643.2 1729.2
- A.36  1   1643.5 1729.5
- A.26  1   1644.5 1730.5
- A.24  1   1646.1 1732.1
- A.4   1   1646.1 1732.1
- A.33  1   1646.7 1732.7
- A.20  1   1647.0 1733.0
- A.8   1   1648.4 1734.4
- A.6   1   1648.6 1734.6
- A.29  1   1649.1 1735.1
- A.48  1   1652.2 1738.2
- A.49  1   1652.2 1738.2
- A.17  1   1654.2 1740.2
- A.44  1   1654.5 1740.5
- A.41  1   1657.9 1743.9
- A.57  1   1661.8 1747.8
- A.21  1   1662.8 1748.8
- A.52  1   1665.6 1751.6
- A.56  1   1669.2 1755.2
- A.23  1   1669.7 1755.7
- A.5   1   1674.9 1760.9
- A.42  1   1679.5 1765.5
- A.45  1   

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Step:  AIC=1723.97
spam ~ A.1 + A.2 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + A.10 + 
    A.12 + A.15 + A.16 + A.17 + A.19 + A.20 + A.21 + A.23 + A.24 + 
    A.25 + A.26 + A.27 + A.28 + A.29 + A.33 + A.35 + A.36 + A.38 + 
    A.39 + A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + A.47 + A.48 + 
    A.49 + A.52 + A.53 + A.54 + A.56 + A.57



Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"Warning mess

       Df Deviance    AIC
<none>      1638.0 1724.0
- A.10  1   1640.6 1724.6
- A.15  1   1640.8 1724.8
- A.9   1   1640.8 1724.8
- A.47  1   1642.1 1726.1
- A.43  1   1642.1 1726.1
- A.38  1   1642.2 1726.2
- A.12  1   1642.6 1726.6
- A.19  1   1643.8 1727.8
- A.39  1   1644.1 1728.1
- A.35  1   1644.3 1728.3
- A.28  1   1644.8 1728.8
- A.2   1   1645.0 1729.0
- A.1   1   1645.2 1729.2
- A.36  1   1645.3 1729.3
- A.26  1   1646.2 1730.2
- A.24  1   1647.8 1731.8
- A.4   1   1647.9 1731.9
- A.33  1   1648.9 1732.9
- A.20  1   1650.0 1734.0
- A.8   1   1650.5 1734.5
- A.6   1   1650.5 1734.5
- A.29  1   1651.0 1735.0
- A.49  1   1654.4 1738.4
- A.48  1   1654.5 1738.5
- A.17  1   1656.2 1740.2
- A.54  1   1657.5 1741.5
- A.44  1   1658.5 1742.5
- A.41  1   1660.0 1744.0
- A.57  1   1664.0 1748.0
- A.21  1   1664.8 1748.8
- A.52  1   1667.5 1751.5
- A.23  1   1671.6 1755.6
- A.56  1   1671.9 1755.9
- A.5   1   1676.5 1760.5
- A.42  1   1683.2 1767.2
- A.45  1   1688.3 1772.3
- A.46  1   


Call:  glm(formula = spam ~ A.1 + A.2 + A.4 + A.5 + A.6 + A.7 + A.8 + 
    A.9 + A.10 + A.12 + A.15 + A.16 + A.17 + A.19 + A.20 + A.21 + 
    A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + A.33 + A.35 + 
    A.36 + A.38 + A.39 + A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + 
    A.47 + A.48 + A.49 + A.52 + A.53 + A.54 + A.56 + A.57, family = binomial, 
    data = spam.train)

Coefficients:
(Intercept)          A.1          A.2          A.4          A.5          A.6  
  -1.591408    -0.589773    -0.139307     2.804753     0.588000     0.817371  
        A.7          A.8          A.9         A.10         A.12         A.15  
   2.112301     0.476626     0.456388     0.113649    -0.159631     1.054460  
       A.16         A.17         A.19         A.20         A.21         A.23  
   1.136034     0.840119     0.089574     1.152511     0.274152     2.062435  
       A.24         A.25         A.26         A.27         A.28         A.29  
   0.369408    -1.970884    -1.071402   -10.680024     

## Wine

In [5]:
nrow(wine)
wine

[1] 178

Type Alcohol Malic Ash  Alcalinity Magnesium Phenols Flavonoids
1   1    14.23   1.71  2.43 15.6       127       2.80    3.06      
2   1    13.20   1.78  2.14 11.2       100       2.65    2.76      
3   1    13.16   2.36  2.67 18.6       101       2.80    3.24      
4   1    14.37   1.95  2.50 16.8       113       3.85    3.49      
5   1    13.24   2.59  2.87 21.0       118       2.80    2.69      
6   1    14.20   1.76  2.45 15.2       112       3.27    3.39      
7   1    14.39   1.87  2.45 14.6        96       2.50    2.52      
8   1    14.06   2.15  2.61 17.6       121       2.60    2.51      
9   1    14.83   1.64  2.17 14.0        97       2.80    2.98      
10  1    13.86   1.35  2.27 16.0        98       2.98    3.15      
11  1    14.10   2.16  2.30 18.0       105       2.95    3.32      
12  1    14.12   1.48  2.32 16.8        95       2.20    2.43      
13  1    13.75   1.73  2.41 16.0        89       2.60    2.76      
14  1    14.75   1.73  2.39 11.4        91       3.10    3.69      
15  1    14.38   1.87  2.38 12.0       102       3.30    3.64      
16  1    13.63   1.81  2.70 17.2       112       2.85    2.91      
17  1    14.30   1.92  2.72 20.0       120       2.80    3.14      
18  1    13.83   1.57  2.62 20.0       115       2.95    3.40      
19  1    14.19   1.59  2.48 16.5       108       3.30    3.93      
20  1    13.64   3.10  2.56 15.2       116       2.70    3.03      
21  1    14.06   1.63  2.28 16.0       126       3.00    3.17      
22  1    12.93   3.80  2.65 18.6       102       2.41    2.41      
23  1    13.71   1.86  2.36 16.6       101       2.61    2.88      
24  1    12.85   1.60  2.52 17.8        95       2.48    2.37      
25  1    13.50   1.81  2.61 20.0        96       2.53    2.61      
26  1    13.05   2.05  3.22 25.0       124       2.63    2.68      
27  1    13.39   1.77  2.62 16.1        93       2.85    2.94      
28  1    13.30   1.72  2.14 17.0        94       2.40    2.19      
29  1    13.87   1.90  2.80 19.4       107       2.95    2.97      
30  1    14.02   1.68  2.21 16.0        96       2.65    2.33      
... ...  ...     ...   ...  ...        ...       ...     ...       
149 3    13.32   3.24  2.38 21.5        92       1.93    0.76      
150 3    13.08   3.90  2.36 21.5       113       1.41    1.39      
151 3    13.50   3.12  2.62 24.0       123       1.40    1.57      
152 3    12.79   2.67  2.48 22.0       112       1.48    1.36      
153 3    13.11   1.90  2.75 25.5       116       2.20    1.28      
154 3    13.23   3.30  2.28 18.5        98       1.80    0.83      
155 3    12.58   1.29  2.10 20.0       103       1.48    0.58      
156 3    13.17   5.19  2.32 22.0        93       1.74    0.63      
157 3    13.84   4.12  2.38 19.5        89       1.80    0.83      
158 3    12.45   3.03  2.64 27.0        97       1.90    0.58      
159 3    14.34   1.68  2.70 25.0        98       2.80    1.31      
160 3    13.48   1.67  2.64 22.5        89       2.60    1.10      
161 3    12.36   3.83  2.38 21.0        88       2.30    0.92      
162 3    13.69   3.26  2.54 20.0       107       1.83    0.56      
163 3    12.85   3.27  2.58 22.0       106       1.65    0.60      
164 3    12.96   3.45  2.35 18.5       106       1.39    0.70      
165 3    13.78   2.76  2.30 22.0        90       1.35    0.68      
166 3    13.73   4.36  2.26 22.5        88       1.28    0.47      
167 3    13.45   3.70  2.60 23.0       111       1.70    0.92      
168 3    12.82   3.37  2.30 19.5        88       1.48    0.66      
169 3    13.58   2.58  2.69 24.5       105       1.55    0.84      
170 3    13.40   4.60  2.86 25.0       112       1.98    0.96      
171 3    12.20   3.03  2.32 19.0        96       1.25    0.49      
172 3    12.77   2.39  2.28 19.5        86       1.39    0.51      
173 3    14.16   2.51  2.48 20.0        91       1.68    0.70      
174 3    13.71   5.65  2.45 20.5        95       1.68    0.61      
175 3    13.40   3.91  2.48 23.0       102       1.80    0.7

vEiem que hi ha molt pocs necessitarem fer algun metode de resampling com boostrap


## Discriminador lineal de Fisher

In [17]:
library(MASS)
wine

Type Alcohol Malic Ash  Alcalinity Magnesium Phenols Flavonoids
1   1    14.23   1.71  2.43 15.6       127       2.80    3.06      
2   1    13.20   1.78  2.14 11.2       100       2.65    2.76      
3   1    13.16   2.36  2.67 18.6       101       2.80    3.24      
4   1    14.37   1.95  2.50 16.8       113       3.85    3.49      
5   1    13.24   2.59  2.87 21.0       118       2.80    2.69      
6   1    14.20   1.76  2.45 15.2       112       3.27    3.39      
7   1    14.39   1.87  2.45 14.6        96       2.50    2.52      
8   1    14.06   2.15  2.61 17.6       121       2.60    2.51      
9   1    14.83   1.64  2.17 14.0        97       2.80    2.98      
10  1    13.86   1.35  2.27 16.0        98       2.98    3.15      
11  1    14.10   2.16  2.30 18.0       105       2.95    3.32      
12  1    14.12   1.48  2.32 16.8        95       2.20    2.43      
13  1    13.75   1.73  2.41 16.0        89       2.60    2.76      
14  1    14.75   1.73  2.39 11.4        91       3.10    3.69      
15  1    14.38   1.87  2.38 12.0       102       3.30    3.64      
16  1    13.63   1.81  2.70 17.2       112       2.85    2.91      
17  1    14.30   1.92  2.72 20.0       120       2.80    3.14      
18  1    13.83   1.57  2.62 20.0       115       2.95    3.40      
19  1    14.19   1.59  2.48 16.5       108       3.30    3.93      
20  1    13.64   3.10  2.56 15.2       116       2.70    3.03      
21  1    14.06   1.63  2.28 16.0       126       3.00    3.17      
22  1    12.93   3.80  2.65 18.6       102       2.41    2.41      
23  1    13.71   1.86  2.36 16.6       101       2.61    2.88      
24  1    12.85   1.60  2.52 17.8        95       2.48    2.37      
25  1    13.50   1.81  2.61 20.0        96       2.53    2.61      
26  1    13.05   2.05  3.22 25.0       124       2.63    2.68      
27  1    13.39   1.77  2.62 16.1        93       2.85    2.94      
28  1    13.30   1.72  2.14 17.0        94       2.40    2.19      
29  1    13.87   1.90  2.80 19.4       107       2.95    2.97      
30  1    14.02   1.68  2.21 16.0        96       2.65    2.33      
... ...  ...     ...   ...  ...        ...       ...     ...       
149 3    13.32   3.24  2.38 21.5        92       1.93    0.76      
150 3    13.08   3.90  2.36 21.5       113       1.41    1.39      
151 3    13.50   3.12  2.62 24.0       123       1.40    1.57      
152 3    12.79   2.67  2.48 22.0       112       1.48    1.36      
153 3    13.11   1.90  2.75 25.5       116       2.20    1.28      
154 3    13.23   3.30  2.28 18.5        98       1.80    0.83      
155 3    12.58   1.29  2.10 20.0       103       1.48    0.58      
156 3    13.17   5.19  2.32 22.0        93       1.74    0.63      
157 3    13.84   4.12  2.38 19.5        89       1.80    0.83      
158 3    12.45   3.03  2.64 27.0        97       1.90    0.58      
159 3    14.34   1.68  2.70 25.0        98       2.80    1.31      
160 3    13.48   1.67  2.64 22.5        89       2.60    1.10      
161 3    12.36   3.83  2.38 21.0        88       2.30    0.92      
162 3    13.69   3.26  2.54 20.0       107       1.83    0.56      
163 3    12.85   3.27  2.58 22.0       106       1.65    0.60      
164 3    12.96   3.45  2.35 18.5       106       1.39    0.70      
165 3    13.78   2.76  2.30 22.0        90       1.35    0.68      
166 3    13.73   4.36  2.26 22.5        88       1.28    0.47      
167 3    13.45   3.70  2.60 23.0       111       1.70    0.92      
168 3    12.82   3.37  2.30 19.5        88       1.48    0.66      
169 3    13.58   2.58  2.69 24.5       105       1.55    0.84      
170 3    13.40   4.60  2.86 25.0       112       1.98    0.96      
171 3    12.20   3.03  2.32 19.0        96       1.25    0.49      
172 3    12.77   2.39  2.28 19.5        86       1.39    0.51      
173 3    14.16   2.51  2.48 20.0        91       1.68    0.70      
174 3    13.71   5.65  2.45 20.5        95       1.68    0.61      
175 3    13.40   3.91  2.48 23.0       102       1.80    0.7

In [31]:
for (i in 1:10){
    n<- nrow(wine)
    iTrain<-sample(1:n,n,replace = TRUE)
    wine.train<-wine[iTrain,]
    wine.test<-wine[-iTrain,]
    wine.lda1<-lda(Type~.,data=wine.train)
    C<-table("True"=wine.test$Type,"Predicted"=wine.pred$class)
    cat("\n length iTest ", nrow(wine.test),'\n')
    print(C)
    cat("\n")
}

ERROR: Error in table(True = wine.test$Type, Predicted = wine.pred$class): all arguments must have the same length
